File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
def dfs(graph, start, visited):
    stack = [start]
    
    while stack:
        for neighbor in graph[stack.pop()]:
            if neighbor not in visited:
                stack.append(neighbor)
                visited.add(neighbor)

def solution(n, computers):
    graph = [set() for _ in range(len(computers))]
    for i, c in enumerate(computers):
        for j, neighbor in enumerate(c):
            if i == j: continue
            if neighbor == 0: continue
            graph[i].add(j)
    
    visited = set()
    count = 0
    for i, node in enumerate(graph):
        if i in visited: continue
        dfs(graph, i, visited)
        count += 1
    return count

File: 프로그래머스/lv3/43162. 네트워크/네트워크.js
// computers에 담긴 연결 따라 연결 시켜주고
// dfs 돌리면서 같은 네트워크 체크 cnt += 1
// return cnt
function solution(n, computers) {
    function dfs(start) {
        const stack = [start]
        visited.add(start)
        
        while (stack.length) {
            const curr = stack.pop()
            
            for (let i = 0; i < n; i++) {
                if (visited.has(i) || graph[curr][i] === 0 || curr === i) continue
                stack.push(i)
                visited.add(i)
            }
        }
    }
    
    const graph = Array(n).fill().map((_, i) => computers[i])
    const visited = new Set()
    // const visited = Array(n).fill().map(() => new Set())
    // visited.forEach((s, i) => s.add(i))
    
    let answer = 0
    for (let i = 0; i < n; i++) {
        if (visited.has(i)) continue
        dfs(i)
        answer += 1
    }
    return answer
}